### PR TITLE
Feature/basic fuota

### LINKED ISSUE #69: basic FUOTA
**Summary:** 

A FUOTA feature will be a great addition to the set of available applications.

**Why do we need this?**

A critical system and LoRaWAN feature that will allow for the adaption of the device and showcases the potential of LoRaWAN.

**What is already there? What do you see now?**

- Example implementation for the older generic node.
- ST FUOTA demo 

**What is missing? What do you want to see?**

An implementation of a device specific FUOTA demo.

**How do you propose to implement this?**

- [x] Implement a basic bootloader
- [x] Implement a LoRaWAN node update client
- [x] Implement a server side LoRaWAN update client

**Environment:**

BareMetal

**Acceptance Criteria:**

- Compatible with LoRaWAN specs

**What can you do yourself and what do you need help with?**

All

### LINKED ISSUE #82: LoRaWAN update client
**Summary:** 

Implement FUOTA update client that supports LoRaWAN multicast and fragmentation specifications.
Part of #69 

**Why do we need this?**

To be able to do LoRaWAN FUOTA

**What is already there? What do you see now?**

Examples using Mbed OS and LMN

**What is missing? What do you want to see?**

A baremetal implementation that allows devices to receive the fragmented data and stores them in the external flash

**How do you propose to implement this?**

- Use latest LoRaWAN specifications
- Implement a standardized interfaces similar to bluetooth OTA

**Environment:**

Baremetal 

**Acceptance Criteria:**

- Functions according to specifications and allows for interoperability

**What can you do yourself and what do you need help with?**

All

### LINKED ISSUE #196: Broken multicast support (McRootKey is always generated based on LoRaWAN version 1.1.0 scheme)
**Summary:** 

Our current LoRaWAN stack generates the MC keys based on 1.1 even when configured for 1.0.3

This is a mirror of an already fixed issue on LMN version 4.4.5

https://github.com/Lora-net/LoRaMac-node/issues/936

### PR DESCRIPTION
**Pull request by elsalahy**
**Summary:**

Adds basic fuota app and server
The application attempts to handle an interoperability fuota test.
closes #69 
closes #82 
closes #196 

**Changes:**

- A number of changes to the stack and the LmHandler
- Adds a basic_fuota app

**Notes for Reviewers:**

Please observe the video I shared on slack.
@johanstokking do you think I should add the JS test server within this application, or just point to [this repo](https://github.com/elsalahy/test-fuota-server) in the documentation.

**Release Notes: _(optional)_**

...

### CODE REVIEW

**Comment by johanstokking** [comment]
> @johanstokking do you think I should add the JS test server within this application, or just point to [this repo](https://github.com/elsalahy/test-fuota-server) in the documentation.

Pointing to the repo is fine with me, thanks.

Anything specific I should look at here?

**Comment by elsalahy** [comment]
@johanstokking

> Pointing to the repo is fine with me, thanks.

Good input, will update the documentation accordingly after your review.

> Anything specific I should look at here?

Yes, only two things:

- A Quick look at the stack changes/patches done, with commits that have `lib` prefix in them
- Check overall application flow (using video in slack HW channel) and functionality in this [lora_app.c](https://github.com/TheThingsIndustries/generic-node-se/blob/101f0840517453c0d10d31b38c5cf633578f7419/Software/app/basic_fuota/lora_app.c)

and that is it from your side, the rest is typical stuff.

**Review by johanstokking — DISMISSED** [review]

**Comment by elsalahy** [comment]
This pull request closes #196 as well

File: Software/app/basic_fuota/lora_app.c [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+    if (IsMcSessionStarted == false)    /* we are in Class A*/
+    {
+      if (IsClockSynched == false)    /* we request AppTimeReq to allow FUOTA */
+      {
+        APP_PPRINTF("\r\n Clock sync in progress, requesting AppTimeReq to intiate FUOTA \r\n");
+        status = LmhpClockSyncAppTimeReq();
+      }
+      else
+      {
+        APP_PPRINTF("\r\n Clock sync successfull, sending random uplink to continue with FUOTA \r\n");
```
Comment (marnixcro) [inline Software/app/basic_fuota/lora_app.c:299]: Spelling: successful

File: Software/app/basic_fuota/lora_app.c [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+  IsTxFramePending = 0;
+  CRITICAL_SECTION_END();
+  if (isPending == 1)
+  {
+    if (LmHandlerIsBusy() == true)
+    {
+      return;
+    }
+
+    if (IsMcSessionStarted == false)    /* we are in Class A*/
```
Comment (marnixcro) [inline Software/app/basic_fuota/lora_app.c:290]: Nitpicking: addressing as 'we' is kinda weird (happens a few times, but only in this function). Maybe change this to more formal language

File: Software/app/basic_fuota/lora_app.c [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+    }
+    else  /* Now we are in Class C or in Class B -- FUOTA feature could be activated */
+    {
+      if (IsFileTransferDone == false)
+      {
+        /* do nothing up to the transfer done or sent a data user */
+      }
+      else
+      {
+        APP_PPRINTF("\r\n File transfare successfull, sending the CRC32 value of the received file\r\n");
```
Comment (marnixcro) [inline Software/app/basic_fuota/lora_app.c:319]: Spelling: transfer, successful

File: Software/app/basic_lorawan/.project [hunk 4]
```diff
@@ -1442,6 +1442,31 @@
 			<type>1</type>
 			<locationURI>PARENT-2-PROJECT_LOC/target/stm32wlxx_hal_driver/Inc/Legacy/stm32_hal_legacy.h</locationURI>
 		</link>
+		<link>
+			<name>lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/packages/FragDecoder.c</name>
+			<type>1</type>
+			<locationURI>PARENT-2-PROJECT_LOC/lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/Packages/FragDecoder.c</locationURI>
```
Comment (marnixcro) [inline Software/app/basic_lorawan/.project:1448]: 'Packages' in the locationURI section of each of these should be 'packages' as STM32CubeIDE will not build. This is also for all the instances below in this file.

File: Software/app/secure_element_lorawan/.project [hunk 5]
```diff
@@ -1517,6 +1517,31 @@
 			<type>1</type>
 			<locationURI>PARENT-2-PROJECT_LOC/target/stm32wlxx_hal_driver/Inc/Legacy/stm32_hal_legacy.h</locationURI>
 		</link>
+		<link>
+			<name>lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/packages/FragDecoder.c</name>
+			<type>1</type>
+			<locationURI>PARENT-2-PROJECT_LOC/lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/Packages/FragDecoder.c</locationURI>
```
Comment (marnixcro) [inline Software/app/secure_element_lorawan/.project:1523]: 'Packages' in the locationURI section of each of these should be 'packages' as STM32CubeIDE will not build. This is also for all the instances below in this file

File: Software/app/sensors_lorawan/.project [hunk 6]
```diff
@@ -1442,6 +1442,31 @@
 			<type>1</type>
 			<locationURI>PARENT-2-PROJECT_LOC/target/stm32wlxx_hal_driver/Inc/Legacy/stm32_hal_legacy.h</locationURI>
 		</link>
+		<link>
+			<name>lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/packages/FragDecoder.c</name>
+			<type>1</type>
+			<locationURI>PARENT-2-PROJECT_LOC/lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/Packages/FragDecoder.c</locationURI>
```
Comment (marnixcro) [inline Software/app/sensors_lorawan/.project:1448]: 'Packages' in the locationURI section of each of these should be 'packages' as STM32CubeIDE will not build. This is also for all the instances below in this file

File: Software/app/basic_fuota/.project [hunk 7]
```diff
@@ -0,0 +1,1631 @@
...
+		</link>
+		<link>
+			<name>target/stm32wlxx_hal_driver/Inc/Legacy/stm32_hal_legacy.h</name>
+			<type>1</type>
+			<locationURI>PARENT-2-PROJECT_LOC/target/stm32wlxx_hal_driver/Inc/Legacy/stm32_hal_legacy.h</locationURI>
+		</link>
+		<link>
+			<name>lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/packages/FragDecoder.c</name>
+			<type>1</type>
+			<locationURI>PARENT-2-PROJECT_LOC/lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/Packages/FragDecoder.c</locationURI>
```
Comment (marnixcro) [inline Software/app/basic_fuota/.project:1448]: 'Packages' in the locationURI section of each of these should be 'packages' as STM32CubeIDE will not build. This is also for all the instances below in this file

File: Software/app/freefall_lorawan/.project [hunk 8]
```diff
@@ -1442,6 +1442,31 @@
 			<type>1</type>
 			<locationURI>PARENT-2-PROJECT_LOC/target/stm32wlxx_hal_driver/Inc/Legacy/stm32_hal_legacy.h</locationURI>
 		</link>
+		<link>
+			<name>lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/packages/FragDecoder.c</name>
+			<type>1</type>
+			<locationURI>PARENT-2-PROJECT_LOC/lib/STM32WLxx_LoRaWAN/LoRaWAN/LmHandler/Packages/FragDecoder.c</locationURI>
```
Comment (marnixcro) [inline Software/app/freefall_lorawan/.project:1448]: 'Packages' in the locationURI section of each of these should be 'packages' as STM32CubeIDE will not build. This is also for all the instances below in this file

**Review by marnixcro — CHANGES_REQUESTED** [review]
This seems to work like in the video, although STM32CubeIDE will not build due to the capitalization issues listed below.

File: Software/app/basic_fuota/lora_app.c (hunk 2 above)
Comment (elsalahy) [reply Software/app/basic_fuota/lora_app.c:290]: Sure

File: Software/app/basic_fuota/lora_app.c (hunk 3 above)
Comment (elsalahy) [reply Software/app/basic_fuota/lora_app.c:319]: fixed

File: Software/app/basic_lorawan/.project (hunk 4 above)
Comment (elsalahy) [reply Software/app/basic_lorawan/.project]: fixed on all files

File: Software/app/basic_fuota/lora_app.c [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+
+    if (IsMcSessionStarted == false)
+    {
+      /*
+       * Currently in Class A
+       * Request AppTimeReq to initiate FUOTA
+       */
+      if (IsClockSynched == false)
+      {
+        APP_PPRINTF("\r\n Clock sync in progress, requesting AppTimeReq to intiate FUOTA \r\n");
```
Comment (marnixcro) [inline Software/app/basic_fuota/lora_app.c:298]: Spelling: initiate

**Review by marnixcro — APPROVED** [review]
All STM32CubeIDE apps compile. Did find a small spelling mistake, you can fix if you still want to.

File: Software/app/basic_fuota/lora_app.c (hunk 9 above)
Comment (elsalahy) [reply Software/app/basic_fuota/lora_app.c]: fixed

**Review by marnixcro — APPROVED** [review]

**Pull request merged**